modules/cef/event-format-parser.c

```diff
@@ -0,0 +1,323 @@
+  FilterXObject *dict_val = filterx_string_new(value, value_len);
+
+  gboolean ok = filterx_object_set_subscript(out, dict_key, &dict_val);
+
+  filterx_object_unref(dict_key);
+  filterx_object_unref(dict_val);
+  return ok;
+}
+
+gboolean unescape_value_separators(KVScanner *self)
+{
+  gchar escaped_separator[3] = {'\\', self->value_separator, '\0'};
+
+  const gchar *start = self->value->str;
+  const gchar *pos = start;
+
+  while ((pos = g_strstr_len(start, self->value->len - (pos - start), escaped_separator)) != NULL)
+    {
+      g_string_append_len(self->decoded_value, start, pos - start);
+      g_string_append_c(self->decoded_value, self->value_separator);
+      start = pos + 2;
+    }
+  g_string_append(self->decoded_value, start);
+  return TRUE;
+}
+
+
+FilterXObject *
+parse_extensions(FilterXFunctionEventFormatParser *self, const gchar *input, GError **error, gpointer user_data)
+{
+  FilterXObject *fillable = (FilterXObject *)user_data;
+  FilterXObject *output = filterx_object_create_dict(fillable);
+
+  KVScanner kv_scanner;
+  kv_scanner_init(&kv_scanner, self->config.extensions_value_separator, self->config.extensions_pair_separator, FALSE);
+  kv_scanner_set_transform_value(&kv_scanner, unescape_value_separators);
+  kv_scanner_input(&kv_scanner, input);
+  while (kv_scanner_scan_next(&kv_scanner))
+    {
+      const gchar *name = kv_scanner_get_current_key(&kv_scanner);
+      gsize name_len = kv_scanner_get_current_key_len(&kv_scanner);
+      const gchar *value = kv_scanner_get_current_value(&kv_scanner);
+      gsize value_len = kv_scanner_get_current_value_len(&kv_scanner);
+
+      if (!_set_dict_value(output, name, name_len, value, value_len))
+        goto exit;
+    }
+
+exit:
+  kv_scanner_deinit(&kv_scanner);
+  return output;
+}
+
+static inline gboolean
+_fill_object_col(FilterXFunctionEventFormatParser *self, gint64 index, const gchar *input, FilterXObject *fillable,
+                 GError **error)
+{
+  Field f = field(self, index);
+  FilterXObject *key = filterx_string_new(f.name, -1);
+  FilterXObject *val = NULL;
+
+  if (!f.field_parser)
+    val = parse_default(self, input, error, fillable);
+  else
+    val = f.field_parser(self, input, error, fillable);
+
+  gboolean ok = FALSE;
+  if (!*error)
+    ok = filterx_object_set_subscript(fillable, key, &val);
+
+  filterx_object_unref(val);
+  filterx_object_unref(key);
+  return ok;
+}
+
+gboolean
+parse(FilterXFunctionEventFormatParser *self, const gchar *log, FilterXObject *fillable, GError **error)
+{
+  gboolean ok = FALSE;
+  gsize num_fields = self->config.num_fields;
+
+  CSVScanner csv_scanner;
+  csv_scanner_init(&csv_scanner, &self->csv_opts, log);
+
+  guint64 i = 0;
+  while (csv_scanner_scan_next(&csv_scanner))
+    {
+      if (i >= num_fields)
+        break;
+
+      const gchar *input = csv_scanner_get_current_value(&csv_scanner);
+      APPEND_ZERO(input, input, csv_scanner_get_current_value_len(&csv_scanner));
```

Comment:
Why don't we pass the `len` down the call chain? We could save a lot of `memcpy`s caused by `APPEND_ZERO` and some `strlen`s down the line.